packages/block-editor/src/components/writing-flow/index.js

```diff
@@ -209,13 +216,55 @@ export default function WritingFlow( { children } ) {
 		hasMultiSelection,
 		blocks,
 		isNavigationMode,
-	} = useSelect( selector );
+		isSelectionEnabled,
+		blockSelectionStart,
+	} = useSelect( selector, [] );
 	const {
 		multiSelect,
 		selectBlock,
 		clearSelectedBlock,
+		setNavigationMode,
 	} = useDispatch( 'core/block-editor' );
 
+	function onMouseDown( event ) {
+		verticalRect.current = null;
+
+		// Clicking inside a selected block should exit navigation mode.
+		if (
+			isNavigationMode &&
+			selectedBlockClientId &&
+			isInsideRootBlock( getBlockDOMNode( selectedBlockClientId ), event.target )
+		) {
+			setNavigationMode( false );
+		}
+
+		// Multi-select blocks when Shift+clicking.
+		if (
+			isSelectionEnabled &&
+			// The main button.
+			// https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
+			event.button === 0
+		) {
+			const clientId = getBlockClientId( event.target );
+
+			if ( clientId ) {
+				if ( event.shiftKey ) {
+					if ( blockSelectionStart !== clientId ) {
+						multiSelect( blockSelectionStart, clientId );
+						event.preventDefault();
+					}
+				// Allow user to escape out of a multi-selection to a singular
+				// selection of a block via click. This is handled here since
+				// onFocus excludes blocks involved in a multiselection, as
```

Comment:
Nit: This was copied from BlockListBlock, but since "onFocus" doesn't exist in this file, it's a bit confusing to read now.